### PR TITLE
[interpreter] implement constrained.callvirt

### PR DESCRIPTION
**Pull request by lewurm**
that's where we are right now:

```
Test run: image=/Users/bernhardu/work/mono/mono/mini/basic.exe
Results: total tests: 134, all pass
Elapsed time: 0.001516 secs (0.001516, 0.000000)

Test run: image=/Users/bernhardu/work/mono/mono/mini/basic-float.exe
Results: total tests: 54, all pass
Elapsed time: 0.000553 secs (0.000553, 0.000000)

Test run: image=/Users/bernhardu/work/mono/mono/mini/basic-long.exe
Results: total tests: 97, all pass
Elapsed time: 0.000755 secs (0.000755, 0.000000)

Test run: image=/Users/bernhardu/work/mono/mono/mini/basic-calls.exe
Results: total tests: 27, all pass
Elapsed time: 0.000437 secs (0.000437, 0.000000)

Test run: image=/Users/bernhardu/work/mono/mono/mini/generics.exe
skip test_0_generic_get_value_optimization_vtype...
skip test_0_ldvirtftn_generic_method...
skip test_0_generic_virtual_on_interfaces...
skip test_0_generic_virtual_on_interfaces_ref...
skip test_0_regress_550964_constrained_enum_long...
skip test_0_fullaot_linq...
skip test_0_fullaot_array_wrappers...
skip test_0_gshared_delegate_rgctx...
skip test_0_gshared_delegate_from_gshared...
skip test_0_bug_620864...
skip test_6_partial_sharing_linq...
skip test_0_partial_shared_method_in_nonshared_class...
skip test_0_regress_668095_synchronized_gshared...
skip test_0_synchronized_gshared...
skip test_0_fullaot_sflda_cctor...
skip test_0_fullaot_imt...
skip test_0_fullaot_variant_iface...
skip test_1_regress_constrained_iface_call_7571...
skip test_0_regress_constrained_iface_call_enum...
skip test_0_delegate_callvirt_fullaot...
skip test_0_constrained_partial_sharing...
skip test_0_delegate_8_args...
Results: total tests: 49, all pass
Elapsed time: 0.020156 secs (0.020156, 0.000000)

Test run: image=/Users/bernhardu/work/mono/mono/mini/objects.exe
skip test_0_unbox_trampoline2...
skip test_0_multi_array_cast...
skip test_42_vtype_delegate...
skip test_42_vtype_virtual_delegate...
skip test_1_store_decimal...
skip test_0_array_get_set_soft_float...
skip test_0_synchronized...
skip test_0_delegate_to_virtual_generic_on_ifaces...
skip test_0_enum_gethashcode_opt...
Results: total tests: 80, all pass
Elapsed time: 0.004027 secs (0.004027, 0.000000)

Overall results: tests: 441, 100% pass
```

and hello world works too :)
```
$ cat /tmp/hello.cs
using System;

public class Sup
{
        public static void Main (string []args) {
                Console.WriteLine ("hello runtime folks");
        }
}
$ ./mono-sgen --interpreter /tmp/hello.exe
hello runtime folks
```

@vargaz I reorganized some stuff around header files and exported symbols according to your comments in https://github.com/mono/mono/commit/defa64af030e9814de385a834f669d17cba2bd90, can you review that part please? (last commit in the PR)

@monojenkins build interpreter

### CODE REVIEW

**Comment by lewurm** [comment]
build interpreter

File: mono/mini/interpreter/interp.c [hunk 1]
```diff
@@ -766,8 +759,8 @@ static MethodArguments* build_args_from_sig (MonoMethodSignature *sig, MonoInvoc
 	if (margs->ilen > 0)
 		margs->iargs = g_malloc0 (sizeof (gpointer) * margs->ilen);
 
-	if (margs->ilen > 3)
-		g_error ("build_args_from_sig: TODO, more than two iregs: %d\n", margs->ilen);
+	if (margs->ilen > 6)
```
Comment (joncham) [inline mono/mini/interpreter/interp.c]: Seems like the check and the error message don't match.

File: mono/mini/tramp-amd64.c [hunk 2]
```diff
@@ -29,6 +29,10 @@
 #include "mini-amd64.h"
 #include "debugger-agent.h"
 
+#ifdef ENABLE_INTERPRETER
+#include "interpreter/interp.h"
+#endif
+
```
Comment (vargaz) [inline mono/mini/tramp-amd64.c]: Why is this needed ?

**Comment by vargaz** [comment]
The interp.h changes look ok. I assume interp-internals.h is the 'internal' definitions shared by the files in interpreter/ ? Also, it would have been better to name the directory 'interp', its easier to type :).

File: mono/mini/interpreter/interp.c (hunk 1 above)
Comment (lewurm) [inline mono/mini/interpreter/interp.c]: thanks! updated :)

File: mono/mini/tramp-amd64.c (hunk 2 above)
Comment (lewurm) [inline mono/mini/tramp-amd64.c]: because `mono_arch_get_enter_icall_trampoline` is defined in `interp.h`, but implemented in `tramp-$arch.c`.  maybe I should move that definition to somewhere else? `mini.h`?

Comment (vargaz) [inline mono/mini/tramp-amd64.c]: Yeah, all the other mono_arch_ functions are defined there in one block.

**Comment by lewurm** [comment]
@vargaz yes, it's meant to be "internal".  Only functions defined in `interp.h` should be used outside of the interpreter.  I'll rename the directory in a future PR 😉

File: mono/mini/tramp-amd64.c (hunk 2 above)
Comment (lewurm) [inline mono/mini/tramp-amd64.c]: okay, added another commit: https://github.com/mono/mono/pull/4359/commits/9e8ab7366d632f485f333f8856a77474400e3c2a
